### DISCUSSION TITLE
Why do we use the htmlspecialchars function in the clean method of the Request class?

### DISCUSSION THREAD

**Discussion by emomaliev** [opencart/opencart#10466] [General]
(https://github.com/opencart/opencart/blob/91a3ece2c4a9616240b1d88001c12eb24f770529/upload/system/library/request.php#L47
 
**htmlspecialchars** is an output filter, to be used only when actually outputting data into a HTML-aware environment.  It's not an input filter. I see this function as useless in this case and potentially dangerous as it can corrupt the data. Perhaps I do not know the reason for such an implementation, could you explain?

**Comment by ADDCreative** [comment]
It is used as an escape function for outputs. It's just applied to the inputs. If you removed it there would be no escaping of the output. If you need the un-escaped version you would need to decode and then escape as necessary for any output.

**Reply by emomaliev** [reply]
Why forcefully escape $_POST, $_GET..? We use the twig template engine, which has escaping enabled.

**Reply by emomaliev** [reply]
Perhaps this was necessary in version 2 of the opentact with the previous templateisotor, but now this line and the method itself are useless in my opinion.

**Reply by ADDCreative** [reply]
> Why forcefully escape $_POST, $_GET..? We use the twig template engine, which has escaping enabled.

Escaping is disable. Probable because of support for other template engines.
https://github.com/opencart/opencart/blob/a8b6e9f1537d4a15199fc2f06812e23c6787fb13/upload/system/library/template/twig.php#L93

**Comment by mhcwebdesign** [comment]
You should always know what you output, and not rely on the template engine to do all the modifications of the output data. There are themes and/or OpenCart extensions which may use a different template engine. In fact, Twig is one of the worst template engines around, it just happens to be widely used.
